dev-docs/RFCs/v1.0/generic-edit-mode.md

```diff
@@ -0,0 +1,100 @@
+
+## API
+
+The `EditMode` interface serves as the core abstraction to editing using nebula.gl. It uses a reactive style approach using callbacks to notify of changes to state and reacting to state changes via exposing an `updateState` function.
+
+```javascript
+
+export type ModeState<TData, TGuides> = {
+  // The data being edited, this can be an array or an object
+  data: TData,
+
+  // Additional configuration for this mode
+  modeConfig: any,
+
+  // The indexes of the selected features
+  selectedIndexes: number[],
+
+  // Features that can be used as a guide for editing the data
+  guides: ?TGuides,
+
+  // The cursor type, as a [CSS Cursor](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor)
+  cursor: string,
+
+  // Callback used to notify applications of an edit action
+  onEdit: (editAction: EditAction<TData>) => void,
+
+  // Callback used to update guides
+  onUpdateGuides: (guides: ?TGuides) => void,
+
+  // Callback used to update cursor
+  onUpdateCursor: (cursor: string) => void
+};
+
+export interface EditMode<TData, TGuides> {
+  // Called every time something in `state` changes
+  updateState(state: ModeState<TData, TGuides>): void;
+
+  // Called when the pointer went down and up without dragging regardless of whether something was picked
+  handleClick(event: ClickEvent): void;
+
+  // Called when the pointer moved, regardless of whether the pointer is down, up, and whether something was picked
+  handlePointerMove(event: PointerMoveEvent): void;
+
+  // Called when the pointer went down on something rendered by this layer and the pointer started to move
+  handleStartDragging(event: StartDraggingEvent): void;
+
+  // Called when the pointer went down on something rendered by this layer, the pointer moved, and now the pointer is up
+  handleStopDragging(event: StopDraggingEvent): void;
+}
+```
+
+A user of this interface will need to call `updateState` anytime the data within `ModeState` change. This is similar to how React calls `render` and how deck.gl calls `Layer.updateState` function.
+
+An implementation of a mode is intended to override the `handle...` functions in order to handle user input. The mode can then call the callbacks provided in `ModeState` (e.g. `onEdit`).
+
+## Integration to react-map-gl-draw
+
+TODO
+
+## Integration with nebula
+
+### Module layout
+
+We will need a `@nebula.gl/core` module separate from `nebula.gl` module. The reason is because this new `@nebula/core` should have no deck.gl dependency.
+
+* `nebula.gl`
+  * depends on `@nebula.gl/core`, `@nebula.gl/layers`, and all the other `@nebula/...` modules.
+  * doesn't have anything in it, just basically imports from the others and re-exports them
+* `@nebula.gl/core`
+  * no (large) dependencies, ideally no dependencies
+  * contains `EditMode` class
+  * contains other general purpose types and classes (e.g. event types like `ClickEvent`)
+* `@nebula.gl/layers`
```

Comment:
When reading this, `layers` is a little non-descript. 

For better reading maybe call this `editable-layers`?

The idea is that this would be a new deck.gl layer catalog (just hosted in a separate library).

```
import * from '@deck.gl/layers';
import * from '@nebula.gl/editable-layers';

Reply:
We'll take this as a separate activity. `@nebula.gl/layers` has been the name for some time.